File: Sources/OpenVPN/CryptoOpenSSL_ObjC/bridged/CryptoCTR.m
//
//  CryptoCTR.m
//  Partout
//
//  Created by Davide De Rosa on 6/15/25.
//  Copyright (c) 2025 Davide De Rosa. All rights reserved.
//
//  https://github.com/passepartoutvpn
//
//  This file is part of Partout.
//
//  Partout is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  Partout is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with Partout.  If not, see <http://www.gnu.org/licenses/>.
//

#import "CryptoBridging.h"
#import "CryptoCTR.h"
#import "ZeroingData.h"
#import "crypto_ctr.h"

@interface CryptoCTR () {
    crypto_ctr_t *ptr;
}
@end

@implementation CryptoCTR

- (nullable instancetype)initWithCipherName:(NSString *)cipherName
                                 digestName:(NSString *)digestName
                                  tagLength:(NSInteger)tagLength
                              payloadLength:(NSInteger)payloadLength
                                      error:(NSError **)error
{
    NSLog(@"PartoutOpenVPN: Using CryptoCTR (bridged ObjC/C)");
    NSParameterAssert(cipherName && [[cipherName uppercaseString] hasSuffix:@"CTR"]);
    NSParameterAssert(digestName);

    self = [super init];
    if (self) {
        ptr = crypto_ctr_create([cipherName cStringUsingEncoding:NSUTF8StringEncoding],
                                [digestName cStringUsingEncoding:NSUTF8StringEncoding],
                                tagLength,
                                payloadLength);
        if (!ptr) {
            return nil;
        }
    }
    return self;
}

- (void)dealloc
{
    crypto_ctr_free(ptr);
}

- (int)cipherIVLength
{
    return (int)ptr->cipher_iv_len;
}

- (int)digestLength
{
    return (int)ptr->meta.digest_length;
}

- (int)tagLength
{
    return (int)ptr->meta.tag_length;
}

- (NSInteger)encryptionCapacityWithLength:(NSInteger)length
{
    return ptr->meta.encryption_capacity(ptr, length);
}

#pragma mark Encrypter

- (void)configureEncryptionWithCipherKey:(ZeroingData *)cipherKey hmacKey:(ZeroingData *)hmacKey
{
    ptr->encrypter.configure(ptr, cipherKey.ptr, hmacKey.ptr);
}

- (BOOL)encryptBytes:(const uint8_t *)bytes length:(NSInteger)length dest:(uint8_t *)dest destLength:(NSInteger *)destLength flags:(const CryptoFlags * _Nullable)flags error:(NSError * _Nullable __autoreleasing * _Nullable)error
{
    crypto_flags_t cf = crypto_flags_from(flags);
    crypto_error_t code;
    if (!ptr->encrypter.encrypt(ptr, bytes, length, dest, (size_t *)destLength, &cf, &code)) {
        if (error) {
            *error = [NSError errorWithDomain:PartoutCryptoErrorDomain code:code userInfo:nil];
        }
        return NO;
    }
    return YES;
}

#pragma mark Decrypter

- (void)configureDecryptionWithCipherKey:(ZeroingData *)cipherKey hmacKey:(ZeroingData *)hmacKey
{
    ptr->decrypter.configure(ptr, cipherKey.ptr, hmacKey.ptr);
}

- (BOOL)decryptBytes:(const uint8_t *)bytes length:(NSInteger)length dest:(uint8_t *)dest destLength:(NSInteger *)destLength flags:(const CryptoFlags * _Nullable)flags error:(NSError * _Nullable __autoreleasing * _Nullable)error
{
    crypto_flags_t cf = crypto_flags_from(flags);
    crypto_error_t code;
    if (!ptr->decrypter.decrypt(ptr, bytes, length, dest, (size_t *)destLength, &cf, &code)) {
        if (error) {
            *error = [NSError errorWithDomain:PartoutCryptoErrorDomain code:code userInfo:nil];
        }
        return NO;
    }
    return YES;
}

- (BOOL)verifyBytes:(const uint8_t *)bytes length:(NSInteger)length flags:(const CryptoFlags * _Nullable)flags error:(NSError * _Nullable __autoreleasing * _Nullable)error
{
    [NSException raise:NSInvalidArgumentException format:@"Verification not supported"];
    return NO;
}

@end

File: Sources/OpenVPN/CryptoOpenSSL/Native/CryptoCTR.swift
//
//  CCryptoCTR.swift
//  Partout
//
//  Created by Davide De Rosa on 6/14/25.
//  Copyright (c) 2025 Davide De Rosa. All rights reserved.
//
//  https://github.com/passepartoutvpn
//
//  This file is part of Partout.
//
//  Partout is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  Partout is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with Partout.  If not, see <http://www.gnu.org/licenses/>.
//

internal import _PartoutCryptoOpenSSL_C
import Foundation

public final class CryptoCTR: Encrypter, Decrypter {
    private let ptr: UnsafeMutablePointer<crypto_ctr_t>

    private let mappedError: (CryptoError) -> Error

    public init(
        cipherName: String,
        digestName: String,
        tagLength: Int,
        payloadLength: Int,
        mappedError: ((CryptoError) -> Error)? = nil
    ) throws {
        guard let ptr = crypto_ctr_create(cipherName, digestName, tagLength, payloadLength) else {
            throw CryptoError()
        }
        NSLog("PartoutOpenVPN: Using CryptoCTR (Swift)")
        self.ptr = ptr
        self.mappedError = mappedError ?? { $0 }
    }

    public var digestLength: Int {
        ptr.pointee.crypto.meta.digest_length
    }

    public var tagLength: Int {
        ptr.pointee.crypto.meta.tag_length
    }

    public func encryptionCapacity(for length: Int) -> Int {
        ptr.pointee.crypto.meta.encryption_capacity(ptr, length)
    }

    public func configureEncryption(withCipherKey cipherKey: ZeroingData?, hmacKey: ZeroingData?) {
        guard let cipherKey, let hmacKey else {
            return
        }
        ptr.pointee.crypto.encrypter.configure(ptr, cipherKey.ptr, hmacKey.ptr)
    }

    public func encryptBytes(_ bytes: UnsafePointer<UInt8>, length: Int, dest: UnsafeMutablePointer<UInt8>, destLength: UnsafeMutablePointer<Int>, flags: CryptoFlagsWrapper?) throws -> Bool {
        var code = CryptoErrorGeneric
        var cFlags = crypto_flags_t()
        let flagsPtr = flags.pointer(to: &cFlags)
        guard ptr.pointee.crypto.encrypter.encrypt(ptr, dest, destLength, bytes, length, flagsPtr, &code) else {
            throw mappedError(CryptoError(code))
        }
        return true
    }

    public func configureDecryption(withCipherKey cipherKey: ZeroingData?, hmacKey: ZeroingData?) {
        guard let cipherKey, let hmacKey else {
            return
        }
        ptr.pointee.crypto.decrypter.configure(ptr, cipherKey.ptr, hmacKey.ptr)
    }

    public func decryptBytes(_ bytes: UnsafePointer<UInt8>, length: Int, dest: UnsafeMutablePointer<UInt8>, destLength: UnsafeMutablePointer<Int>, flags: CryptoFlagsWrapper?) throws -> Bool {
        var code = CryptoErrorGeneric
        var cFlags = crypto_flags_t()
        let flagsPtr = flags.pointer(to: &cFlags)
        guard ptr.pointee.crypto.decrypter.decrypt(ptr, dest, destLength, bytes, length, flagsPtr, &code) else {
            throw mappedError(CryptoError(code))
        }
        return true
    }

    public func verifyBytes(_ bytes: UnsafePointer<UInt8>, length: Int, flags: CryptoFlagsWrapper?) throws -> Bool {
        fatalError("Unsupported")
    }
}
